### ISSUE TITLE
Question regarding reserved category, HINT instructions and illegal exception 

### ISSUE THREAD

**Issue by taoj33** [riscv/riscv-isa-manual#300]
Hi,
In SPEC it defines Reserved as
 > Reserved encoding are currently not defined but are saved for future standard extensions

while for HINT, it is defined as following
> HINTS do not change any architecture visible state,except for advancing the pc and any applicable performance counters.

Is it correct to interpret that HINT is a subset of Reserved category which can be treated as NOP while  all other encodings in the Reserved category should be treated as illegal instructions?

**Comment by aswaterman** [comment]
HINTs are not in the Reserved opcode category; they live within an already-defined standard encoding space.

You are right that HINTs should be treated as NOPs, whereas Reserved opcodes should raise illegal-instruction exceptions on current conforming implementations.

**Comment by aswaterman** [comment]
-

**Comment by taoj33** [comment]
@aswaterman  Thank you for your reply.

I can understand instructions with invalid func/opcode should trigger traps. 

But should instructions with undefined imm/reg raise illegal traps? for example, 
> C.JR  (RES, rs1=0)
> SLLIW,SRIW, and SRAIW encodings with imm!=0 are reserved

**Comment by aswaterman** [comment]
Ah.  I understand the confusion, as the C chapter is written a bit differently and needs to be cleaned up.  The two examples you quoted are in different classes:

- The shift instructions with imm=0 are marked HINT in Tables 16.6 and 16.7, so they should execute as NOPs (which they do naturally, since 0 is the shift identity).
- C.JR with rs1=0 is marked RES, rather than HINT, so should raise an exception.

**Comment by taoj33** [comment]
@aswaterman 
Sry made a typo 

> SLLIW, SRLIW, and SRAIW encodings with $imm[5] !=0 are reserved.

> Previously, SLLIW, SRLIW, and SRAIW with imm[5]=0 were defined to cause illegal instruction 
> exceptions, whereas now they are marked as reserved.  This is a backwards-compatible change.

So is it safe to conclude all instructions described as reserved should raise exception?

**Comment by aswaterman** [comment]
Yes.  The difference between "is reserved" and "should raise an exception" is that, in the former case, you can put a non-compliant extension in that encoding space.  If you aren't using that space for a non-compliant extension, you should simply raise an exception.

**Comment by asb** [comment]
Where in the spec does it say that instructions in reserved encoding spaces should trap? That's obviously a sensible and desirable thing to do, but I can't see the spec requiring it.

**Comment by aswaterman** [comment]
Yeah, that’s why I said “should”; it’s not required. However, reserved encodings could eventually be used for new standard purposes. So implementations that don’t trap them could eventually become non-conforming to newer versions of the spec, since they are incapable of emulating the new instructions.

**Comment by asb** [comment]
Thanks for the clarification, wasn't sure if you were using an IETF style SHOULD.
